components/common-go/grpc/ratelimit.go

```diff
@@ -6,60 +6,201 @@ package grpc
 
 import (
 	"context"
+	"strings"
 	"time"
 
 	"golang.org/x/time/rate"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
 
 	"github.com/gitpod-io/gitpod/common-go/util"
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/prometheus/client_golang/prometheus"
 )
 
+type keyFunc func(req interface{}) (string, error)
+
 // RateLimit configures the reate limit for a function
 type RateLimit struct {
 	Block          bool          `json:"block"`
 	BucketSize     uint          `json:"bucketSize"`
 	RefillInterval util.Duration `json:"refillInterval"`
+
+	KeyCacheSize uint   `json:"keyCacheSize,omitempty"`
+	Key          string `json:"key,omitempty"`
+}
+
+func (r RateLimit) Limiter() *rate.Limiter {
+	return rate.NewLimiter(rate.Every(time.Duration(r.RefillInterval)), int(r.BucketSize))
 }
 
 // NewRatelimitingInterceptor creates a new rate limiting interceptor
 func NewRatelimitingInterceptor(f map[string]RateLimit) RatelimitingInterceptor {
-	funcs := make(map[string]ratelimitedFunction, len(f))
+	callCounter := prometheus.NewCounterVec(prometheus.CounterOpts{
```

Comment:
I believe the metrics here also need to be registered, to be picked up by prometheus, otherwise they are just local counters which won't show up on `/metrics` endpoint.

See https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#Register

Reply:
In this case, I'd actually hoist the counter definition outside of this instantiation and make it package private variable. You can use https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/promauto to simplify the registration.

Reply:
My bad, missed this. Normally, I'd expect the registration to happen in this file and automatically without requiring an extra call. The risk here is that a component uses the rate limiting interceptor but doesn't register the metric.

The other issue is that this would panic on startup, but CI doesn't catch this.

Reply:
For a while now we've adopted the pattern to make things implement the `prometheus.Collector` interface so that they can be registered.
Testing is just as easy as passing in the registry, but it doesn't mandate that someone uses metrics.
The "registration error at runtime" behaviour is also the same.

Reply:
Thanks. I'd personally like the need for registration to be more explicit (and to use DI to pass the registry in) but given this is an existing pattern, better to stick to it than introduce yet another. Thanks for the info.